### PR TITLE
Implements the feedback from PerRowVersioning PR

### PR DESCRIPTION
**Pull request by dakota002**
@lpsinger  This method could probably be refactored a bit, but this would allow for the implementation as you described in #16

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -60,85 +64,244 @@ export interface DynamoDBAutoIncrementProps {
...
+            await this.props.doc.query({
+              KeyConditionExpression: 'widgetID = :widgetID',
+              ExpressionAttributeValues: {
+                ':widgetID': 1,
+              },
+              TableName: this.props.counterTableName,
+            })
+          ).Items
+          console.log(`History: ${JSON.stringify(temp, null, 2)}`)
+          console.log(Put)
```
Comment (lpsinger) [inline src/index.ts:169]:
```suggestion
```

File: src/index.ts [hunk 2]
```diff
@@ -60,85 +64,244 @@ export interface DynamoDBAutoIncrementProps {
...
 
-      const Put: PutCommandInput = {
-        ConditionExpression: 'attribute_not_exists(#counter)',
-        ExpressionAttributeNames: { '#counter': this.props.tableAttributeName },
-        Item: { [this.props.tableAttributeName]: nextCounter, ...item },
-        TableName: this.props.tableName,
+        if (this.props.dangerously) {
+          const temp = (
+            await this.props.doc.query({
+              KeyConditionExpression: 'widgetID = :widgetID',
```
Comment (lpsinger) [inline src/index.ts:161]: What's with the attribute name `widgetID` hardcoded here?

File: src/index.ts [hunk 3]
```diff
@@ -60,85 +64,244 @@ export interface DynamoDBAutoIncrementProps {
...
 
-      let nextCounter: number
-      let Update: UpdateCommandInput & { UpdateExpression: string }
+    if (this.props.counterTableCopyItem) {
+      const ExpressionAttributeNames = this.#getExpressionAttributeNames(item)
+      const ExpressionAttributeValues = this.#getExpressionAttributeValues(item)
+      const UpdateExpression = this.#getUpdateExpression(item)
+      for (;;) {
+        const previousItem = await this.getLastItem()
+        const counter = previousItem?.[this.#attributeName]
```
Comment (lpsinger) [inline src/index.ts:105]: Why the new method? What is wrong with `this.getLast()`?

**Review by lpsinger — CHANGES_REQUESTED** [review]
- There is now a lot of duplicated logic in this function: the `if (this.props.counterTableCopyItem)` and `else` branches are extremely similar. Please refactor to reduce duplication. (The only difference between the two branches should be that the Update expression has some extra keys and values.)
- There is new code to escape every single one of the keys in the item to be put with `ExpressionAttributeName`. That's a nice feature, but it seems orthogonal to the goal of the PR. Please split those changes out.

File: src/index.ts (hunk 2 above)
Comment (dakota002) [reply src/index.ts]: this was from a log test, removed it in the latest commit

File: src/index.ts (hunk 3 above)
Comment (dakota002) [reply src/index.ts:105]: `getLast` returns the latest value of the `#attributeName` from the table defined as `counterTableName`

`getLastItem` returns the latest value of the item defined on `tableName`, which is later PUT into the table defined as `counterTableName` 

The example case of circulars: 
  - `getLast` will return the latest circularId
  - `getLastItem` returns the latest circular

**Comment by lpsinger** [comment]
> * There is now a lot of duplicated logic in this function: the `if (this.props.counterTableCopyItem)` and `else` branches are extremely similar. Please refactor to reduce duplication. (The only difference between the two branches should be that the Update expression has some extra keys and values.)

It might simplify the logic to turn the `Update` command into a `Put` command. I think that `Put` doesn't require field names in the item to be escaped whereas `Update` does.

**Comment by dakota002** [comment]
> > * There is now a lot of duplicated logic in this function: the `if (this.props.counterTableCopyItem)` and `else` branches are extremely similar. Please refactor to reduce duplication. (The only difference between the two branches should be that the Update expression has some extra keys and values.)
> 
> It might simplify the logic to turn the `Update` command into a `Put` command. I think that `Put` doesn't require field names in the item to be escaped whereas `Update` does.

I can look into that, if I remember correctly, the `Put` command Up-serts into the table, so it should work out

**Pull request closed without merging**